File: Obj-C/NativeAd/NativeAd/TableViewController.m
#import "TableViewController.h"
#import "BDMContentTableViewCell.h"
#import "BDMNativeAdViewCell.h"

#import <BidMachine/BidMachine.h>

@interface RefreshControll : UIRefreshControl

@end

@implementation RefreshControll

- (instancetype)init {
    if (self = [super init]) {
        self.attributedTitle = [[NSAttributedString alloc] initWithString: @"Load ad"];
    }
    return self;
}

@end

@interface TableViewController ()<BidMachineAdDelegate>

@property (nonatomic, assign) NSInteger cellsCount;

@property (nonatomic, strong) BidMachineNative *nativeAd;

@end

@implementation TableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.tableView registerNib:BDMContentTableViewCell.nib forCellReuseIdentifier:BDMContentTableViewCell.reuseIdentifier];
    [self.tableView registerNib:BDMNativeAdViewCell.nib forCellReuseIdentifier:BDMNativeAdViewCell.reuseIdentifier];
    self.tableView.refreshControl = [RefreshControll new];
    [self.tableView.refreshControl addTarget:self action:@selector(updateData) forControlEvents:UIControlEventValueChanged];
}


- (void)updateData {
    NSError *error = nil;
    id<BidMachineRequestConfigurationProtocol> configuration = [BidMachineSdk.shared requestConfiguration: BidMachinePlacementFormatNative error:&error];
    if (error != nil) {
        [self.refreshControl endRefreshing];
        return;
    }
    
    __weak __typeof(self) weakSelf = self;
    [BidMachineSdk.shared native:configuration :^(BidMachineNative *ad, NSError *error) {
        if (ad == nil) {
            [self.refreshControl endRefreshing];
            return;
        }
        
        weakSelf.nativeAd = ad;
        weakSelf.nativeAd.controller = weakSelf;
        weakSelf.nativeAd.delegate = weakSelf;
        [weakSelf.nativeAd loadAd];
    }];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    if ((indexPath.row % 20) == 0) {
        BDMNativeAdViewCell *nativeAdViewCell = [tableView dequeueReusableCellWithIdentifier:BDMNativeAdViewCell.reuseIdentifier forIndexPath:indexPath];
        [self.nativeAd unregisterView];
        [self.nativeAd presentAd:nativeAdViewCell :nativeAdViewCell error:nil];
        return nativeAdViewCell;
    }
    
    BDMContentTableViewCell *fakeCell = [tableView dequeueReusableCellWithIdentifier:BDMContentTableViewCell.reuseIdentifier forIndexPath:indexPath];
    fakeCell.fakeTitle.text = @"Lorem ipsum";
    fakeCell.fakeDescription.text = @"Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.";
    UInt32 rand = arc4random_uniform(2);
    if ((rand % 2) == 0) {
        fakeCell.fakeIcon.image = [UIImage imageNamed:@"ManPic"];
    } else {
        fakeCell.fakeIcon.image = [UIImage imageNamed:@"KittyPic"];
    }
    
    return fakeCell;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.cellsCount;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    if ((indexPath.row % 20) == 0) {
        return 375;
    }
    return 86;
}

#pragma mark - BidMachineAdDelegate

- (void)didLoadAd:(id<BidMachineAdProtocol> _Nonnull)ad {
    [self.tableView.refreshControl endRefreshing];
    self.cellsCount = 100;
    [self.tableView reloadData];
}

- (void)didFailLoadAd:(id<BidMachineAdProtocol> _Nonnull)ad :(NSError * _Nonnull)error {
    [self.tableView.refreshControl endRefreshing];
    if (self.cellsCount == 0) {
        return;
    }
    self.cellsCount = 0;
    [self.tableView reloadData];
}

@end

File: Swift/NativeAd/NativeAd/TableViewController.swift
import UIKit
import BidMachine

class RefreshControll: UIRefreshControl {
    
    override init() {
        super.init()
        self.attributedTitle = NSAttributedString(string: "Load ad")
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

class TableViewController: UITableViewController {
    
    struct TableViewContentEntity {
        var image: UIImage
        var title: String
        var description: String
    }
    
    private struct Constants {
        static let contentTitle = "Lorem ipsum"
        static let contentDescription = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."
    }
    
    private var cellsCount = 0
    
    private var nativeAd: BidMachineNative?
    
    private var fakeEntityArray: [TableViewContentEntity] = [
        TableViewContentEntity(image: UIImage(named: "ManPic")!, title: Constants.contentTitle, description: Constants.contentDescription),
        TableViewContentEntity(image: UIImage(named: "KittyPic")!, title: Constants.contentTitle, description: Constants.contentDescription),
        TableViewContentEntity(image: UIImage(named: "ManPic")!, title: Constants.contentTitle, description: Constants.contentDescription)
    ]
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.tableView.register(ContentTableViewCell.nib, forCellReuseIdentifier: ContentTableViewCell.reuseIdentifier)
        self.tableView.register(NativeAdViewCell.nib, forCellReuseIdentifier: NativeAdViewCell.reuseIdentifier)
        self.tableView.refreshControl = RefreshControll()
        self.tableView.refreshControl?.addTarget(self, action: #selector(updataData(_:)), for: .valueChanged)
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return cellsCount
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        if indexPath.row.quotientAndRemainder(dividingBy: 20).remainder == 0 {
            let nativeAdViewCell = tableView.dequeueReusableCell(withIdentifier: NativeAdViewCell.reuseIdentifier, for: indexPath) as! NativeAdViewCell
            nativeAd?.unregisterView()
            try? nativeAd?.presentAd(nativeAdViewCell, nativeAdViewCell)
            return nativeAdViewCell
        }
        
        let fakeCell = tableView.dequeueReusableCell(withIdentifier: ContentTableViewCell.reuseIdentifier, for: indexPath) as! ContentTableViewCell
        
        let entity = fakeEntityArray[Int.random(in: 0...2)]
        
        fakeCell.fakeIcon.image = entity.image
        fakeCell.fakeTitle.text = entity.title
        fakeCell.fakeDescription.text = entity.description
        
        return fakeCell
    }
    
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        if indexPath.row.quotientAndRemainder(dividingBy: 20).remainder == 0 {
            return 375
        }
        return 86
    }
}

extension TableViewController: BidMachineAdDelegate {
    
    func didLoadAd(_ ad: BidMachine.BidMachineAdProtocol) {
        self.tableView.refreshControl?.endRefreshing()
        cellsCount = 100
        tableView.reloadData()
    }
    
    func didFailLoadAd(_ ad: BidMachine.BidMachineAdProtocol, _ error: Error) {
        self.tableView.refreshControl?.endRefreshing()
        guard cellsCount != 0 else {
            return
        }
        cellsCount = 0
        tableView.reloadData()
    }
}

extension TableViewController {
    
    @objc private func updataData(_ sender: Any) {
        BidMachineSdk.shared.native { [weak self] ad, error in
            guard let self = self, let ad = ad else {
                self?.refreshControl?.endRefreshing()
                return
            }
            self.nativeAd = ad
            ad.controller = self
            ad.delegate = self
            ad.loadAd()
        }
    }
    
}
 
